### PR TITLE
add script to restore Java project dependencies by Maven command

### PR DESCRIPTION
**Pull request by anmeng10101**

### CODE REVIEW

File: src/java2yaml/Common/Utility/Guard.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    using System;
+
+    /// <summary>
+    /// Represents a simple class for validating parameters and throwing exceptions.
+    /// </summary>
+    public static class Guard
+    {
+        /// <summary>
+        /// Validates argumentValue is not null and throws ArgumentNullException if it is null.
+        /// n
```
Comment (superyyrrzz) [inline src/java2yaml/Common/Utility/Guard.cs]: remove?

File: src/java2yaml/Program.cs [hunk 2]
```diff
@@ -74,5 +79,20 @@ private static bool ValidateConfig(string[] args)
 
             return true;
         }
+
+        private static bool IsDocletExstis()
+        {
+            var doclet = Path.Combine(PathUtility.GetAssemblyDirectory(), Constants.DocletLocation);
+            
+            if(File.Exists(doclet))
```
Comment (superyyrrzz) [inline src/java2yaml/Program.cs]: minor: add `space` after `if`

File: src/java2yaml/Steps/RestoreDependency.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    using System.Threading.Tasks;
+    using System.Linq;
+
+    public class RestoreDependency : IStep
+    {
+        public RestoreDependency(string path)
+        {
+            RepositoryPath = path;
+        }
+        public string StepName
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/RestoreDependency.cs]: for simplicity: `public string StepName { get; }`

File: src/java2yaml/Steps/GenerateDocument.cs [hunk 4]
```diff
@@ -13,7 +13,22 @@ public Task RunAsync(ConfigModel config)
...
             return Task.Run(() =>
             {
-                //TODO
+                foreach (var path in config.RepositoryFolders)
+                {
+                    var procedure = new StepCollection(
+                        new RestoreDependency(path),
+                        new RunJavadoc(path));
+
+                    try
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/GenerateDocument.cs]: Feels no need to try/catch here. `StepCollection.RunAsync()` already has a try/catch wrapping this.

File: src/java2yaml/Common/Utility/FileUtility.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+            return new DirectoryInfo(rootPath)
+                .GetFiles(fileName, SearchOption.AllDirectories);
+        }
+
+        public static IEnumerable<string> GetFilesByExtension(string rootPath, string extension)
+        {
+            Guard.ArgumentNotNullOrEmpty(rootPath, nameof(rootPath));
+
+            if(string.IsNullOrEmpty(extension))
```
Comment (superyyrrzz) [inline src/java2yaml/Common/Utility/FileUtility.cs]: add space after `if`

File: src/java2yaml/Steps/GenerateDocument.cs [hunk 6]
```diff
@@ -13,7 +13,22 @@ public Task RunAsync(ConfigModel config)
...
+                    {
+                        procedure.RunAsync(config).Wait();
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/GenerateDocument.cs]: Usually, it recommends to use `await xxx` over `xxx.Wait()`

File: src/java2yaml/Steps/RestoreDependency.cs (hunk 3 above)
Comment (superyyrrzz) [inline src/java2yaml/Steps/RestoreDependency.cs]: can simplify: `public string StepName => "RestoreDependency"`

File: src/java2yaml/Steps/RestoreDependency.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        public string RepositoryPath { get; }
+
+
+        public Task RunAsync(ConfigModel config)
+        {
+            return Task.Run(() =>
+            {
+                Guard.ArgumentNotNullOrEmpty(RepositoryPath, nameof(RepositoryPath));
+
+                var projectFiles = FileUtility.GetFilesByName(RepositoryPath, Constants.BuildTool.Maven);
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/RestoreDependency.cs]: Do not reuse `IEnumerable`. Or you can reuse it after `ToList()`.

File: src/java2yaml/Steps/RestoreDependency.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                if (projectFiles.Count() == 0)
+                {
+                    ConsoleLogger.WriteLine(new LogEntry
+                    {
+                        Phase = StepName,
+                        Level = LogLevel.Error,
+                        Message = " The project file pom.xml cannot be found, exiting..."
+                    });
+
+                    Environment.Exit(1);
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/RestoreDependency.cs]: You can use `throw` here. Outside has a catch to handle this. Feels not like a common usage to interrupt inside program.

File: src/java2yaml/Steps/RestoreDependency.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+            string commandLineArgs = "/c mvn dependency:copy-dependencies -DoutputDirectory=" + packagePath;
+
+            ConsoleLogger.WriteLine(new LogEntry
+            {
+                Phase = StepName,
+                Level = LogLevel.Info,
+                Message = $" Restoring dependency under directory '{workingDirectory}'..."
+            });
+
+            ProcessUtility.Execute("cmd.exe", commandLineArgs, workingDirectory);
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/RestoreDependency.cs:59]: Can this be `ProcessUtility.Execute("mvn", commandLineArgs, workingDirectory);`? Feels no need to use `cmd.exe` here.

File: src/java2yaml/Steps/RunJavadoc.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+        public string RepositoryPath { get; }
+
+        public Task RunAsync(ConfigModel config)
+        {
+            return Task.Run(() =>
+            {
+                //TODO
+            }
+            );
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/RunJavadoc.cs:23]: You can put `);` after `}` the line above.

File: src/java2yaml/Steps/RestoreDependency.cs (hunk 9 above)
Comment (anmeng10101) [reply src/java2yaml/Steps/RestoreDependency.cs:59]: Tired to use `ProcessUtility.Execute("mvn.cmd", commandLineArgs, workingDirectory)`, always get exit code 1 even command is correct, need further investigate.

File: src/java2yaml/Steps/GenerateDocument.cs (hunk 4 above)
Comment (anmeng10101) [reply src/java2yaml/Steps/GenerateDocument.cs]: removed unnecessary try/catch

File: src/java2yaml/Common/Utility/Guard.cs (hunk 1 above)
Comment (anmeng10101) [reply src/java2yaml/Common/Utility/Guard.cs]: removed

File: src/java2yaml/Program.cs (hunk 2 above)
Comment (anmeng10101) [reply src/java2yaml/Program.cs]: `space` added

File: src/java2yaml/Common/Utility/FileUtility.cs (hunk 5 above)
Comment (anmeng10101) [reply src/java2yaml/Common/Utility/FileUtility.cs]: `space` added

File: src/java2yaml/Steps/GenerateDocument.cs (hunk 6 above)
Comment (anmeng10101) [reply src/java2yaml/Steps/GenerateDocument.cs]: changed to `await `

File: src/java2yaml/Steps/RestoreDependency.cs (hunk 3 above)
Comment (anmeng10101) [reply src/java2yaml/Steps/RestoreDependency.cs]: corrected

File: src/java2yaml/Steps/RestoreDependency.cs (hunk 8 above)
Comment (anmeng10101) [reply src/java2yaml/Steps/RestoreDependency.cs]: switch to `throw new System.Exception()`

File: src/java2yaml/Steps/RunJavadoc.cs (hunk 10 above)
Comment (anmeng10101) [reply src/java2yaml/Steps/RunJavadoc.cs:23]: corrected

File: src/java2yaml/Steps/RestoreDependency.cs (hunk 7 above)
Comment (anmeng10101) [reply src/java2yaml/Steps/RestoreDependency.cs]: added `ToList()`

File: src/java2yaml/Steps/RestoreDependency.cs [hunk 11]
```diff
@@ -11,40 +11,30 @@ public RestoreDependency(string path)
...
                 {
-                    ConsoleLogger.WriteLine(new LogEntry
-                    {
-                        Phase = StepName,
-                        Level = LogLevel.Error,
-                        Message = " The project file pom.xml cannot be found, exiting..."
-                    });
-
-                    Environment.Exit(1);
+                    throw new System.Exception("POM.xml not available.");
```
Comment (superyyrrzz) [inline src/java2yaml/Steps/RestoreDependency.cs:30]: can use `InvalidOperationException`.

**Review by superyyrrzz — APPROVED** [review]

File: src/java2yaml/Steps/RestoreDependency.cs (hunk 11 above)
Comment (anmeng10101) [reply src/java2yaml/Steps/RestoreDependency.cs:30]: corrected

**Pull request merged**